File: app/src/main/java/com/wolcano/musicplayer/music/mvp/listener/RecyclerViewScrollListener.java
package com.wolcano.musicplayer.music.mvp.listener;


import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public abstract class RecyclerViewScrollListener extends RecyclerView.OnScrollListener {

    private int scrolledDistance = 0;
    private static final int HIDE_THRESHOLD = 20;

    public static int previousTotal = 0;
    private boolean loading = true;
    private int visThreshold = 2;
    int firstVisibleItem, visibleItemCount, totalItemCount;

    private boolean infiniteScrollingEnabled = true;

    private boolean controlsVisible = true;

    protected RecyclerViewScrollListener() {
    }

    @Override
    public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
        super.onScrolled(recyclerView, dx, dy);

        RecyclerView.LayoutManager manager = recyclerView.getLayoutManager();

        visibleItemCount = recyclerView.getChildCount();
        if (manager instanceof GridLayoutManager) {
            GridLayoutManager gridLayoutManager = (GridLayoutManager) manager;
            firstVisibleItem = gridLayoutManager.findFirstVisibleItemPosition();
            totalItemCount = gridLayoutManager.getItemCount();
        } else if (manager instanceof LinearLayoutManager) {
            LinearLayoutManager linearLayoutManager = (LinearLayoutManager) manager;
            firstVisibleItem = linearLayoutManager.findFirstVisibleItemPosition();
            totalItemCount = linearLayoutManager.getItemCount();
        }


        if (infiniteScrollingEnabled) {
            if (loading) {
                if (totalItemCount > previousTotal) {
                    loading = false;
                    previousTotal = totalItemCount;
                }
            }

            if (!loading && (totalItemCount - visibleItemCount <= firstVisibleItem + visThreshold)) {
                onLoadMore();
                loading = true;
            }
        }

        if (firstVisibleItem == 0) {
            if (!controlsVisible) {
                onScrollUp();
                controlsVisible = true;
            }

            return;
        }

        if (scrolledDistance > HIDE_THRESHOLD && controlsVisible) {
            onScrollDown();
            controlsVisible = false;
            scrolledDistance = 0;
        } else if (scrolledDistance < -HIDE_THRESHOLD && !controlsVisible) {
            onScrollUp();
            controlsVisible = true;
            scrolledDistance = 0;
        }

        if ((controlsVisible && dy > 0) || (!controlsVisible && dy < 0)) {
            scrolledDistance += dy;
        }
    }

    public abstract void onScrollUp();

    public abstract void onScrollDown();

    public abstract void onLoadMore();

}

File: app/src/main/java/com/wolcano/musicplayer/music/listener/RecyclerViewScrollListener.kt
package com.wolcano.musicplayer.music.listener

import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

abstract class RecyclerViewScrollListener protected constructor() :
    RecyclerView.OnScrollListener() {
    private var scrolledDistance = 0
    private var loading = true
    var firstVisibleItem = 0
    var visibleItemCount = 0
    var totalItemCount = 0
    private var controlsVisible = true

    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)
        val manager = recyclerView.layoutManager
        visibleItemCount = recyclerView.childCount
        if (manager is GridLayoutManager) {
            firstVisibleItem = manager.findFirstVisibleItemPosition()
            totalItemCount = manager.itemCount
        } else if (manager is LinearLayoutManager) {
            firstVisibleItem = manager.findFirstVisibleItemPosition()
            totalItemCount = manager.itemCount
        }
        val infiniteScrollingEnabled = true
        if (infiniteScrollingEnabled) {
            if (loading) {
                if (totalItemCount > previousTotal) {
                    loading = false
                    previousTotal = totalItemCount
                }
            }
            val visibleThreshold = 2
            if (!loading && totalItemCount - visibleItemCount <= firstVisibleItem + visibleThreshold) {
                onLoadMore()
                loading = true
            }
        }
        if (firstVisibleItem == 0) {
            if (!controlsVisible) {
                onScrollUp()
                controlsVisible = true
            }
            return
        }
        if (scrolledDistance > HIDE_THRESHOLD && controlsVisible) {
            onScrollDown()
            controlsVisible = false
            scrolledDistance = 0
        } else if (scrolledDistance < -HIDE_THRESHOLD && !controlsVisible) {
            onScrollUp()
            controlsVisible = true
            scrolledDistance = 0
        }
        if (controlsVisible && dy > 0 || !controlsVisible && dy < 0) {
            scrolledDistance += dy
        }
    }

    abstract fun onScrollUp()
    abstract fun onScrollDown()
    abstract fun onLoadMore()

    companion object {
        private const val HIDE_THRESHOLD = 20
        var previousTotal = 0
    }

}